### PR TITLE
[TTS] | Wire MCAS | CassandraKvs impl

### PR DESCRIPTION
**Pull request by sudiksha27**
## General
**Before this PR**:
There was no way to do MCAS in AtlasDb

**After this PR**:
Breaks down https://github.com/palantir/atlasdb/pull/6116
Introduces KVS.multiCheckAndSet api

**Priority**:
This week please

**Concerns / possible downsides (what feedback would you like?)**:
Do the docs need to be more elaborate?
In case of failure, is there a subset of columns where update can be successful? It does not appear so taking a cursory look at the code but it will require more testing to gain some confidence.

**Is documentation needed?**:
Docs are added

## Compatibility
**Does this PR create any API breaks (e.g. at the Java or HTTP layers) - if so, do we have compatibility?**:
No

**Does this PR change the persisted format of any data - if so, do we have forward and backward compatibility?**:
No

**The code in this PR may be part of a blue-green deploy. Can upgrades from previous versions safely coexist? (Consider restarts of blue or green nodes.)**:
Not for this change

**Does this PR rely on statements being true about other products at a deployment - if so, do we have correct product dependencies on these products (or other ways of verifying that these statements are true)?**:
No

**Does this PR need a schema migration?**
No

## Testing and Correctness
**What, if any, assumptions are made about the current state of the world? If they change over time, how will we find out?**:
We are assuming that MCAS perf for usage of txn4 will comparable to PUE

**What was existing testing like? What have you done to improve it?**:
Added integration tests

**If this PR contains complex concurrent or asynchronous code, is it correct? The onus is on the PR writer to demonstrate this.**:
Relying on underlying Cassandra impl i.e. if 2 requests try MCAS at the same time only one will succeed.

**If this PR involves acquiring locks or other shared resources, how do we ensure that these are always released?**:
No

## Execution
**How would I tell this PR works in production? (Metrics, logs, etc.)**:
I suppose we could add telemetry around PUE now and then MCAS as well. Although transaction duration should be a good indication of impact of this change.

**Has the safety of all log arguments been decided correctly?**:
N/A

**Will this change significantly affect our spending on metrics or logs?**:
N/A

**How would I tell that this PR does not work in production? (monitors, etc.)**:
This PR actually does not make any change in the current flow

**If this PR does not work as expected, how do I fix that state? Would rollback be straightforward?**:
N/A

**If the above plan is more complex than “recall and rollback”, please tag the support PoC here (if it is the end of the week, tag both the current and next PoC)**:
N/A

## Scale
**Would this PR be expected to pose a risk at scale? Think of the shopping product at our largest stack.**:
Not for this PR

**Would this PR be expected to perform a large number of database calls, and/or expensive database calls (e.g., row range scans, concurrent CAS)?**:
This endpoint will trigger one cas operation on Cassandra.

**Would this PR ever, with time and scale, become the wrong thing to do - and if so, how would we know that we need to do something differently?**:
Not as per our findings from perf testing

## Development Process
**Where should we start reviewing?**:
KeyValueService

**If this PR is in excess of 500 lines excluding versions lock-files, why does it not make sense to split it?**:
The PR is close to 500 lines but ~40% of that is testing.

**Please tag any other people who should be aware of this PR**:
@jeremyk-91
@sverma30
@raiju

### CODE REVIEW

File: atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java [hunk 1]
```diff
@@ -564,6 +566,11 @@ public void checkAndSet(CheckAndSetRequest checkAndSetRequest) throws CheckAndSe
         }
     }
 
+    @Override
+    public void multiCheckAndSet(MultiCheckAndSetRequest multiCheckAndSetRequest) throws MultiCheckAndSetException {
+        throw new UnsupportedOperationException("DbKvs does not support multi-checkAndSet operation!");
+    }
```
Comment (sudiksha27) [inline atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java:572]: Will raise a new PR to deal with this. This is complex enough to be done in a separate PR

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java [hunk 2]
```diff
@@ -494,6 +496,11 @@ public void checkAndSet(CheckAndSetRequest request) throws CheckAndSetException
         }
     }
 
+    @Override
+    public void multiCheckAndSet(MultiCheckAndSetRequest multiCheckAndSetRequest) throws MultiCheckAndSetException {
+        throw new UnsupportedOperationException("InMemoryKvs does not support multi-checkAndSet operation!");
+    }
```
Comment (sudiksha27) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java:502]: Will raise a separate PR for this.

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/DualWriteKeyValueService.java [hunk 3]
```diff
@@ -122,6 +124,11 @@ public void checkAndSet(CheckAndSetRequest checkAndSetRequest) {
         delegate1.checkAndSet(checkAndSetRequest);
     }
 
+    @Override
+    public void multiCheckAndSet(MultiCheckAndSetRequest multiCheckAndSetRequest) throws MultiCheckAndSetException {
+        delegate1.multiCheckAndSet(multiCheckAndSetRequest);
+    }
```
Comment (sudiksha27) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/DualWriteKeyValueService.java:130]: Only used for testing.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java [hunk 4]
```diff
@@ -1982,6 +1984,57 @@ public void checkAndSet(final CheckAndSetRequest request) throws CheckAndSetExce
...
+
+                throw new MultiCheckAndSetException(
+                        LoggingArgs.tableRef(tableRef),
+                        request.rowName(),
+                        request.expected(),
+                        currentValues,
+                        LoggingArgs.isSafe(tableRef));
+            }
+        } catch (MultiCheckAndSetException e) {
+            throw e;
```
Comment (sudiksha27) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceImpl.java:2031]: Following pattern used in this class for consistency.

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/MultiCheckAndSetException.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+package com.palantir.atlasdb.keyvalue.api;
+
+import com.palantir.logsafe.Arg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import java.util.Map;
+
+public class MultiCheckAndSetException extends RuntimeException {
+    private final byte[] rowName;
+    private final Map<Cell, byte[]> expectedValue;
```
Comment (gmaretic) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/MultiCheckAndSetException.java:25]: nit: values

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java [hunk 6]
```diff
@@ -514,6 +520,175 @@ public void setOnceTest() {
...
+        keyValueService.multiCheckAndSet(request);
+    }
+
+    @Test
+    public void testMultiCheckAndSetFromValueWhenNoValue() {
+        Map<Cell, byte[]> expected = ImmutableMap.of(TEST_CELL, val(0, 0));
+        Map<Cell, byte[]> updates = ImmutableMap.of(TEST_CELL, val(0, 1));
+        MultiCheckAndSetRequest request =
+                MultiCheckAndSetRequest.multipleCells(TEST_TABLE, TEST_CELL.getRowName(), expected, updates);
+        assertThatThrownBy(() -> keyValueService.multiCheckAndSet(request))
```
Comment (gmaretic) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java:575]: Here and maybe in a couple additional tests: we want to check that the actual values in the exception are correct

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java [hunk 7]
```diff
@@ -514,6 +520,175 @@ public void setOnceTest() {
...
+                TEST_TABLE, nextTestCell.getRowName(), ImmutableMap.of(nextTestCell, val(0, 1))));
+
+        assertThatThrownBy(() -> keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.newCells(
+                        TEST_TABLE, nextTestCell.getRowName(), ImmutableMap.of(nextTestCell, val(0, 2)))))
+                .isInstanceOf(MultiCheckAndSetException.class);
+
+        verifyMultiCheckAndSet(ImmutableMap.of(firstTestCell, val(0, 0), nextTestCell, val(0, 1)));
+    }
+
+    private void verifyMultiCheckAndSet(Map<Cell, byte[]> expectedValues) {
```
Comment (gmaretic) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java:675]: We might as well test a multicas that puts a new cell and updates an existing one -- we won't use this in txn4, but we do want to know that it works

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/MultiCheckAndSetException.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+    public Map<Cell, byte[]> getExpectedValue() {
+        return getValIfSafe(expectedValue);
+    }
+
+    public Map<Cell, byte[]> getActualValues() {
+        return getValIfSafe(actualValues);
+    }
+
+    private <T> T getValIfSafe(T val) {
```
Comment (gmaretic) [inline atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/MultiCheckAndSetException.java:57]: We don't actually need this if we are not logging things, and we do want to have access to actual values even when not safe for logging

File: atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java [hunk 9]
```diff
@@ -380,6 +382,17 @@ public void checkAndSet(CheckAndSetRequest request) {
...
 
+    @Override
+    public void multiCheckAndSet(MultiCheckAndSetRequest request) throws MultiCheckAndSetException {
+        maybeLog(
+                () -> delegate.multiCheckAndSet(request),
+                logCellsAndSize(
+                        "multiCheckAndSet",
+                        request.tableRef(),
+                        request.updates().size(),
+                        byteSize(request.expected()) + byteSize(request.updates())));
```
Comment (gmaretic) [inline atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java:393]: 👍

**Review by gmaretic — COMMENTED** [review]
This is most of the way there, a few nits and would like a few more tests

File: atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/MultiCheckAndSetException.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.Map;
+
+public class MultiCheckAndSetException extends RuntimeException {
+    private final byte[] rowName;
+    private final Map<Cell, byte[]> expectedValue;
+    private final Map<Cell, byte[]> actualValues;
+    private final boolean isSafe;
+
+    public MultiCheckAndSetException(
+            Arg<String> tableReference,
```
[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/MultiCheckAndSetException.java:30]: yeah i think there is the issue of circular deps and i was being lazy in moving LoggingArgs to a common module

[the remark this answers is not in the document]
Comment (sudiksha27) [reply atlasdb-api/src/main/java/com/palantir/atlasdb/keyvalue/api/MultiCheckAndSetException.java:30]: actually this is quite convoluted as safeLoggingArgs depends on things in atlasdb-client  which then has a dep on atlasdb-api and we are basically adding stuff to atlasdb-api :(

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java (hunk 7 above)
Comment (sudiksha27) [reply atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java:675]: sure added an explicit test - I think the thing to _really_ know is that the "check" will only happen on the columns explicitly declared as expected and on all columns, if no columns are declared. I will add this to the doc.

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java [hunk 11]
```diff
@@ -637,9 +662,13 @@ public void testMcasChecksAllColumnsIfExpectedIsEmpty() {
...
-        assertThatThrownBy(() -> keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.newCells(
-                        TEST_TABLE, firstTestCell.getRowName(), ImmutableMap.of(nextTestCell, secondVal))))
-                .isInstanceOf(MultiCheckAndSetException.class);
+        MultiCheckAndSetException ex = assertThrows(
+                MultiCheckAndSetException.class,
+                () -> keyValueService.multiCheckAndSet(MultiCheckAndSetRequest.newCells(
+                        TEST_TABLE, nextTestCell.getRowName(), ImmutableMap.of(nextTestCell, secondVal))));
+
+        assertThat(ex.getExpectedValues()).isEmpty();
+        assertThat(ex.getActualValues()).isNotEmpty();
```
Comment (gmaretic) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraKeyValueServiceIntegrationTest.java:671]: nit: don't be lazy!

**Review by gmaretic — APPROVED** [review]

**Pull request merged**